PhysicsTools/NanoAOD/plugins/ProtonProducer.cc

```diff
@@ -0,0 +1,225 @@
+          if (ref.key() == tr_idx)
+            multiRP_idx = p_idx;
+        }
+      }
+      multiRPProtonIdx.push_back(multiRP_idx);
+    }
+
+    // update proton tables
+    std::unique_ptr<edm::ValueMap<int>> protonRPIdV(new edm::ValueMap<int>());
+    edm::ValueMap<int>::Filler fillerID(*protonRPIdV);
+    fillerID.insert(hRecoProtonsSingleRP, singleRP_RPId.begin(), singleRP_RPId.end());
+    fillerID.fill();
+
+    std::unique_ptr<edm::ValueMap<bool>> singleRP_sector45V(new edm::ValueMap<bool>());
+    edm::ValueMap<bool>::Filler fillersingle45(*singleRP_sector45V);
+    fillersingle45.insert(hRecoProtonsSingleRP, singleRP_sector45.begin(), singleRP_sector45.end());
+    fillersingle45.fill();
+
+    std::unique_ptr<edm::ValueMap<bool>> singleRP_sector56V(new edm::ValueMap<bool>());
+    edm::ValueMap<bool>::Filler fillersingle56(*singleRP_sector56V);
+    fillersingle56.insert(hRecoProtonsSingleRP, singleRP_sector56.begin(), singleRP_sector56.end());
+    fillersingle56.fill();
+
+    std::unique_ptr<edm::ValueMap<bool>> multiRP_sector45V(new edm::ValueMap<bool>());
+    edm::ValueMap<bool>::Filler fillermulti45(*multiRP_sector45V);
+    fillermulti45.insert(hRecoProtonsMultiRP, multiRP_sector45.begin(), multiRP_sector45.end());
+    fillermulti45.fill();
+
+    std::unique_ptr<edm::ValueMap<bool>> multiRP_sector56V(new edm::ValueMap<bool>());
+    edm::ValueMap<bool>::Filler fillermulti56(*multiRP_sector56V);
+    fillermulti56.insert(hRecoProtonsMultiRP, multiRP_sector56.begin(), multiRP_sector56.end());
+    fillermulti56.fill();
+
+    // build track table
+    auto ppsTab = std::make_unique<nanoaod::FlatTable>(trackX.size(), "PPSLocalTrack", false);
+    ppsTab->addColumn<int>("singleRPProtonIdx", singleRPProtonIdx, "local track - proton correspondence");
+    ppsTab->addColumn<int>("multiRPProtonIdx", multiRPProtonIdx, "local track - proton correspondence");
+    ppsTab->addColumn<float>("x", trackX, "local track x", 16);
+    ppsTab->addColumn<float>("xUnc", trackXUnc, "local track x uncertainty", 8);
+    ppsTab->addColumn<float>("y", trackY, "local track y", 13);
+    ppsTab->addColumn<float>("yUnc", trackYUnc, "local track y uncertainty", 8);
+    ppsTab->addColumn<float>("time", trackTime, "local track time", 16);
+    ppsTab->addColumn<float>("timeUnc", trackTimeUnc, "local track time uncertainty", 13);
+    ppsTab->addColumn<int>("decRPId", decRPId, "local track detector dec id");
+    ppsTab->addColumn<int>("numFitPoints", numFitPoints, "number of points used for fit");
+    ppsTab->addColumn<int>("pixelRecoInfo", pixelRecoInfo, "flag if a ROC was shifted by a bunchx");
+    ppsTab->addColumn<float>("normalizedChi2", normalizedChi2, "chi2 over NDF", 8);
+    ppsTab->addColumn<int>("rpType", rpType, "strip=3, pixel=4, diamond=5, timing=6");
+    ppsTab->addColumn<float>("localSlopeX", localSlopeX, "track horizontal angle", 11);
+    ppsTab->addColumn<float>("localSlopeY", localSlopeY, "track vertical angle", 11);
```

Comment:
the precision of these variable should be rationalized.
Can the precision 16 and 13 be lowered to 11 ?

Reply:
@mariadalfonso 
The rationale for choosing the precision of these variables was to be consistent with the precision here: https://github.com/cms-sw/cmssw/blob/master/RecoPPS/Local/plugins/CTPPSLocalTrackLiteProducer.cc#L152-L157

Reply:
@jwill24 @jan-kaspar

at each step of the RECO/AOD --> miniAOD --> nanoAOD  we reduce the precision size of almost everything. so PPS should follow the same.

Reply:
I am not sure I understand the logic... What we did in PPS for miniAOD was to find precision settings such that the additional error due to bit truncation is negligible to other expected uncertainties. So in principle we don't have any room further reductions. If needed, we could revisit the evaluation and check if could make few compromises here and there. But I don't expect big savings (unless we got it wrong before).

We understand the motivation for reducing the nanoAOD size and we try to do our best. For example, we store only the variables which are actually useful for analyses, see different lists in the single-RP and multi-RP tables. This is not done in miniAOD.

Reply:
@jan-kaspar
in nanoAOD usually we store high level objects, not all the inputs that can be useful for calculation a posteriori.

Also try to have a look to the variable stored for other objects, the precision for floats rarely goes beyond 10 !

Reply:
@mariadalfonso what is the exact meaning of the `precision` parameter? Is it the number of mantissa bits preserved?

@jwill24 would you be available to check the precision estimates and see whether we could get closer to 10 please? We can possibly get in touch with @AndreaBellora who did the previous estimation.

Reply:
@jan-kaspar @mariadalfonso 
This is how the precision was calculated previously:

Concerning the track coordinates, we would need a 5 digit mantissa
like AB.CDE. Since in some cases the value can go above 65 (some pots
use the garage position reference frame) we need a maximum mantissa
value of around 70000. The max unsigned integer that can be expressed
with n bits is 2^n -1. Therefore we need log_2(70001) = 16.09 ->16
bits.

Concerning the track angles, the uncertainty can be estimated with
(thanks Jan for your help) sqrt(2)*sensor_resolution/lever_arm. For
pixels this is: sqrt(2)*(150[100]x10^(-6))(2.17x10^(-2)) =
0.0098[0.0065] for x[y] angles. Both angle values lie between -0.3 and
0.3.

We need therefore 4 decimal digits for the mantissa -> maximum integer
value of around (to be very conservative) 5000.

N = log_2(5001) = 12.28-> 13 bits.